alias.c

```diff
@@ -509,17 +503,17 @@ void mutt_alias_delete_reverse(struct Alias *t)
  */
 int mutt_alias_complete(char *s, size_t buflen)
 {
-  struct Alias *a = Aliases;
-  struct Alias *a_list = NULL, *a_cur = NULL;
+  struct Alias *a = NULL, *tmp = NULL;
+  struct AliasList a_list = TAILQ_HEAD_INITIALIZER(a_list);
   char bestname[HUGE_STRING];
 
   if (s[0] != 0) /* avoid empty string as strstr argument */
   {
     memset(bestname, 0, sizeof(bestname));
 
-    while (a)
+    TAILQ_FOREACH(a, &Aliases, entries)
     {
-      if (a->name && strstr(a->name, s) == a->name)
+      if (a->name && a->name[0] == s[0] && strstr(a->name, s) == a->name)
```

Comment:
If `(strstr(a->name, s) == a->name)`, then we know the two strings match at the beginning,
so I'm not sure I see the need for the check `a->name[0] == s[0]`

Reply:
It's a micro-optimization. `strstr(str, sub)` scans full string `str` to locate `sub`. This avoids scanning at all if first characters do not match.

Reply:
How about  `strncmp(a->name, s, strlen(s)) == 0` ?

Reply:
hmm, the `strstr()` implies that `mutt_alias_complete()` might match on any partial string,
but I can only get it to complete if the beginning matches.

Reply:
as gahr points out, `strstr(a->name, s) == a->name` always anchors the expression to the beginning, so `strncmp()` is much better.

Reply:
@flatcap, @gahr done.